dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java

```diff
@@ -133,88 +133,48 @@ public class DoraLoadJob extends AbstractJob<DoraLoadJob.DoraLoadTask> {
     mBandwidth = bandwidth;
     mUsePartialListing = usePartialListing;
     mVerificationEnabled = verificationEnabled;
-    FileSystem fs = FileSystem.Factory.create(Scheduler.getInstance().getFileSystemContext());
-    ListStatusPOptions listOptions = ListStatusPOptions
-        .newBuilder()
-        .setRecursive(true).build();
-    mFs = fs;
-    mFileIterator = Optional.of(new FileListFetcher(fs, mPath, listOptions, false));
-  }
-
-  private static class FileListFetcher implements Iterator<URIStatus> {
-    public static final int PREFETCH_SIZE = 1000;
-    private final FileSystem mFs;
-    private String mPath;
-    ListStatusPOptions mListOption;
-    public LinkedBlockingQueue<URIStatus> mFiles = new LinkedBlockingQueue<>();
-    private boolean mUsePartialListing = false;
-    private AtomicBoolean mIsDone = new AtomicBoolean(false);
-
-    public FileListFetcher(FileSystem fs, String path, ListStatusPOptions listOption,
-                           boolean usePartialListing) {
-      mPath = path;
-      mListOption = listOption;
-      mFs = fs;
-      mUsePartialListing = false; // default to false now
-    }
-
-    private int advance() {
-      if (mIsDone.get()) {
-        return 0;
-      }
-      try {
-        // TODO(lucy) paginate list a PREFETCH_SIZE here.
-        List<URIStatus> uriStatuses = mFs.listStatus(new AlluxioURI(mPath), mListOption);
-        if (uriStatuses == null) {
-          mIsDone.compareAndSet(false, true);
-          return 0;
-        }
-        uriStatuses.forEach(uriStatus -> mFiles.offer(uriStatus));
-        if (!mUsePartialListing) {
-          mIsDone.compareAndSet(false, true);
-        }
-        return uriStatuses.size();
-      } catch (IOException | AlluxioException e) {
-        throw AlluxioRuntimeException.from(e);
-      }
-    }
-
-    @Override
-    public boolean hasNext() {
-      while (!mIsDone.get() && mFiles.size() < PREFETCH_SIZE * 0.2) {
-        if (advance() <= 0) {
-          break;
-        }
-      }
-      return mFiles.peek() != null;
-    }
+    String ufsRoot = Configuration.getString(PropertyKey.DORA_CLIENT_UFS_ROOT);
+    AlluxioURI ufsRootUri = new AlluxioURI(ufsRoot);
+    AlluxioURI ufsSyncRootUri = ufsRootUri.join(path);
+    mUfs = UnderFileSystem.Factory.create(
+        ufsRoot,
+        UnderFileSystemConfiguration.defaults(Configuration.global()));
+    mLoadMetadataOnly = loadMetadataOnly;
+    LOG.info(
+        "DoraLoadJob for {} created. {} workers are active",
+        path, Preconditions.checkNotNull(Scheduler.getInstance()).getActiveWorkers().size());
 
-    @Override
-    public URIStatus next() {
-      while (!mIsDone.get() && mFiles.size() < PREFETCH_SIZE * 0.2) {
-        if (advance() <= 0) {
-          break;
-        }
-      }
-      URIStatus uriStatus = mFiles.poll();
-      if (uriStatus == null) {
-        throw new NoSuchElementException();
+    try {
```

Comment:
do you mind extracting the mUfsStatusIterator to a class like UfsFileIterable so I can adopt this to other jobs as well?

Reply:
i don't quite understand on this... mUfsStatusIterator is obtained by calling ufs listStatusIterable(). The only thing I did in addition to calling the ufs is setting the full ufs path. I doubt if it is really meaningful to extract this 7-8 lines of code.

Reply:
@jja725  think such class could be developed later with elega's mUfs.listStatusIterable inside and apply the async paginating listing we talked about, and it could be replacing ufsfileiterable as well